### PR TITLE
[Feature] add even_size option to split sampler

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
Add an option to ensure all partitions have the same number of samples. It will add extra samples if needed, like what torch DistributedSampler does. 

@rich-junwang

Example:
```
sampler = SplitSampler(len(train_dataset), num_workers, rank, even_size=True)
train_iter = gluon.data.DataLoader(train_dataset, batch_size=batch_size, sampler=sampler)
```

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

cc @dmlc/gluon-nlp-team

### CODE REVIEW

**Comment by mli** [comment]
Job PR-1028/1 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1028/1/index.html

**Comment by rich-junwang** [comment]
So this change still requires that if we split file, we need to split it into equal size i.e. equal number of lines in each smaller file. The change of sampler guarantees equal number of files is seen by each trainer, right?

**Comment by eric-haibin-lin** [comment]
@rich-junwang you're right. This PR does not directly help on file-based DatasetLoader.

**Comment by mli** [comment]
Job PR-1028/2 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1028/2/index.html

File: src/gluonnlp/data/sampler.py [hunk 1]
```diff
@@ -508,29 +508,48 @@ class SplitSampler(Sampler):
...
+            self._start = self._start if self._start < length else length
+            self._end = self._end if self._end <= length else length
+            self._len = part_len
 
     def __iter__(self):
         # Extract examples between `start` and `end`, shuffle and return them.
         indices = list(range(self._start, self._end))
+        if self.even_size and len(indices) < self._len:
+            # guaranteed to have part_len number of samples
+            candidates = list(range(self._len * self.num_parts))
```
Comment (sxjscience) [inline src/gluonnlp/data/sampler.py]: Should we choose `candidates = list(range(self._length))`?

**Comment by mli** [comment]
Job PR-1028/3 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1028/3/index.html

**Comment by mli** [comment]
Job PR-1028/4 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1028/4/index.html

**Pull request merged**